### PR TITLE
Add deploy-rs, tailscale and users file

### LINKED ISSUE #2: Deploy automation
After the initial deployment using `nixos-anywhere`, what's next? How do we deploy updates?

I don't think we should develop this repo inside the target machine. Options:

- Use `deploy-rs` (we can deploy from macOS)
- Use remote nixos-rebuild switch (not sure if it will work from macOS)

### PR DESCRIPTION
**Pull request by faisalsaifii**
Resolves #2

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -3,9 +3,28 @@
...
+            user = "root";
+            path = deploy-rs.lib.x86_64-linux.activate.nixos self.nixosConfigurations.office;
+          };
+        };
+      };
+      apps."x86_64-linux" = {
+        # Deploy
+        default = {
+          type = "app";
+          program = "${deploy-rs.packages."x86_64-linux".deploy-rs}/bin/deploy";
```
Comment (srid) [inline flake.nix:25]: Note: You can also use `lib.getExe` here

File: flake.nix [hunk 2]
```diff
@@ -17,13 +36,16 @@
...
-                # password you can find the changed IP.
-                initialHashedPassword = "";
                 openssh.authorizedKeys.keys = [
                   "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIFN5Ov2zDIG59/DaYKjT0sMWIY15er1DZCT9SIak07vK"
+                  "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIJnwODoZ7J2O+uCkYUpBb45kWxF5FITXJtzyXy02S3Aj"
+                ];
+              };
+              faisal = {
+                openssh.authorizedKeys.keys = [
+                  "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIJnwODoZ7J2O+uCkYUpBb45kWxF5FITXJtzyXy02S3Aj"
```
Comment (srid) [inline flake.nix:46]: This PR does more than adding deploy-rs? If so, either the PR title should be updated and this be split into two different PRs.

Regardless, I think we should refactor the user key management. By looking at line 41 it is not immediately obvious as to whose key that is. You could create a `users.nix` containing an attrset of users and their properties, and that reference that here for legibility.

**Comment by shivaraj-bh** [comment]
I think it will be nice if we also add nixos VM test to make sure certain SSH keys will be present even after deployment. This will prove to be crucial when someone accidentally raises a PR that removes them and after the deployment the maintainers of the system might not have access to the machine anymore.

Considering this machine will be used by a good number of users.

> This PR does more than adding deploy-rs? If so, either the PR title should be updated and this be split into two different PRs.

In that case it will also be good to rename the PR to have a general name that would encompass the following things:

- deploy-rs
- tailscale setup
- test `authorized_keys` in nixosVM

**Comment by shivaraj-bh** [comment]
> I think it will be nice if we also add nixos VM test to make sure certain SSH keys will be present even after deployment.

On a second thought, I don't think this will be essential. If we ever face an issue, maybe worth spending time.

File: flake.nix [hunk 3]
```diff
@@ -3,9 +3,31 @@
...
   };
-  outputs = { nixpkgs, disko, ... }:
+  outputs = { self, nixpkgs, disko, deploy-rs, ... }:
+    let
+      users = import ./nix/users.nix;
+    in
     {
+      deploy.nodes = {
+        nixos = {
+          hostname = "100.92.110.74"; # Tailscale IP
```
Comment (srid) [inline flake.nix:16]: Perhaps also add the Tailscale account identifier, and information about who has the password.

File: nix/users.nix [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+    "shivaraj" = {
+        "pubKey" = "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIFN5Ov2zDIG59/DaYKjT0sMWIY15er1DZCT9SIak07vK";
+    };
+    "faisal" = {
+        "pubKey" = "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIJnwODoZ7J2O+uCkYUpBb45kWxF5FITXJtzyXy02S3Aj";
+    };
```
Comment (srid) [inline nix/users.nix:7]: Double quote not required in keys.

File: flake.nix [hunk 5]
```diff
@@ -15,15 +37,19 @@
               ./disk-config.nix
             ];
             services.openssh.enable = true;
+            services.tailscale.enable = true;
             users.users = {
               root = {
-                # Post-installation, the IP might change if MAC is not the
-                # only identifier used by DHCP server to lease an IP, by setting a
-                # password you can find the changed IP.
-                initialHashedPassword = "";
```
Comment (srid) [inline flake.nix:23]: @shivaraj-bh Why was this removed?

Comment (shivaraj-bh) [reply flake.nix:23]: Only during the installation of the OS one will need it for reasons mentioned above the attribute.
